runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/UserMonitoringInfoToCounterUpdateTransformer.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.fn.control;
+
+import com.google.api.services.dataflow.model.CounterMetadata;
+import com.google.api.services.dataflow.model.CounterStructuredName;
+import com.google.api.services.dataflow.model.CounterStructuredNameAndMetadata;
+import com.google.api.services.dataflow.model.CounterUpdate;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.MonitoringInfo;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.MonitoringInfoSpecs.Enum;
+import org.apache.beam.runners.core.metrics.SpecMonitoringInfoValidator;
+import org.apache.beam.runners.dataflow.worker.DataflowExecutionContext.DataflowStepContext;
+import org.apache.beam.runners.dataflow.worker.MetricsToCounterUpdateConverter.Origin;
+import org.apache.beam.runners.dataflow.worker.counters.DataflowCounterUpdateExtractor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Class for transforming MonitoringInfo's containing User counter values, to relevant CounterUpdate
+ * proto.
+ */
+class UserMonitoringInfoToCounterUpdateTransformer
+    implements MonitoringInfoToCounterUpdateTransformer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BeamFnMapTaskExecutor.class);
+
+  private final Map<String, DataflowStepContext> transformIdMapping;
+
+  private final SpecMonitoringInfoValidator specValidator;
+
+  public UserMonitoringInfoToCounterUpdateTransformer(
+      final SpecMonitoringInfoValidator specMonitoringInfoValidator,
+      final Map<String, DataflowStepContext> transformIdMapping) {
+    this.transformIdMapping = transformIdMapping;
+    this.specValidator = specMonitoringInfoValidator;
+  }
+
+  static final String BEAM_METRICS_USER_PREFIX =
```

Comment:
This is defined in two files now, SimpleMonitoringInfoBuilder. Can we just define it in one place

Reply:
I see it as this value is defined in one place -- beam_fn_api.proto and both, SimpleMonitoringInfoBuilder and UserMonitoringInfoToCounterUpdateTransformer only define convenient accessor to the field.

I think that creating a structure with these fields separate from proto will only bring more inconvenience to the code.

Reply:
Synced offline. BeamUrns extracts value from MonitoringInfoUrns, while UserMonitoringInfoToCounterUpdateTransformer utilizes MonitoringInfoSpecs. Decided to keep this PR as-is and have a separate PR to remove MonitoringInfoUrns enum and utilize MonitoringInfoSpecs as source of truth.